Refactor to remove duplicated logic

chrismytton opened the pull request:
Tidy up the code for generating OCD ids from CSVs so there's less duplication.

chrismytton committed:
Convert \r to \n in area_info.csv

This file was using CR line endings. I've converted this to LF using the
following command:

    ruby -pi -e '$_.gsub!(?\r, ?\n)' area_info.csv

chrismytton committed:
Refactor to remove duplicated logic

Rather than having two similar but not quite the same ways to handle
parsing the CSV this refactoring moves things into a shared function. We
now use a mapping argument to work around the differences in the CSV
structures.

tmtmtmtm commented on Rakefile:
```diff
@@ -14,66 +14,48 @@ task :generate_csv do
-      name: r[:constituency].strip
-    }
-  end
+  ocd_ids = ocd_ids_from_csv(google_csv, region: :arearegion, subregion: :areasub_region, district: :areadistrict, constituency: :areaconstituency) + ocd_ids_from_csv(area_csv)
```
This took a couple of readings to spot what was going on. I'd be tempted to make those mappings an explicit Hash to stand out more, rather than looking like this is taking a long list of arguments.

tmtmtmtm commented on Rakefile:
```diff
@@ -14,66 +14,48 @@ task :generate_csv do
-      id: id_for(region:r[:arearegion], subregion: r[:areasub_region], district: r[:areadistrict], constituency: r[:areaconstituency]),
-      name: r[:areaconstituency]
-    }
+  def ocd_ids_from_csv(csv_data, mapping = {})
```
Perhaps better to call this `raw_csv_data` or something to make it obvious that the function is going to parse it itself?`</nit>`

Some documentation would probably be helpful here too, particularly to explain what way around the mapping works (i.e. whether it's "column_name_in_sheet => column_name_we_wanted" or "column_name_we_want => where_to_find_it_in_sheet")

chrismytton committed:
Make the CSV mapping for the Google CSV more explicit

By pulling this out into a separate variable it's clearer that this is a
mapping hash, rather  than just looking like a long set of arguments.

chrismytton committed:
Fix rake task description

We're now getting area information from multiple sources.

chrismytton committed:
Make it clearer that ocd_ids_from_csv take a raw CSV string

tmtmtmtm commented on Rakefile:
```diff
@@ -14,66 +14,50 @@ task :generate_csv do
+
+      constituency = r[mapping.fetch(:constituency, :constituency)]
+      next unless constituency
+      next if ['-', '_'].include?(constituency)
```
It's not immediately obvious to me what this is doing, and feels like it might be in the wrong place. (i.e. although this isn't quite a completely generic function, it's fairly close to being able to lift it and turn into a standalone class, whereas this doesn't seem like something that would be part of that — though tricky to know as I can't quite tell why it's here…)

chrismytton replied:
Ah, this is because [a couple of rows in the corrections CSV](https://github.com/theyworkforyou/uganda_ocd_ids/blob/1255164b9eacd2ad0a00fc4f91914c2e5714dd3e/area_info.csv#L2) have `-` or `_` in them, so this was just to skip them. I guess the simplest thing to do is to just remove those cells from the CSV. Alternatively we could have an extra argument to say what cell values should be skipped.

tmtmtmtm replied:
Hmm. It's not obvious to me what those are actually signifying. Is that just the equivalent of being blank? If so, and it's something we want to handle more generally, then we could have a more generic "is_empty?" check that looks for those as well as blank. Otherwise I think this should definitely be handled at a different level, which might be easier by splitting this up into two functions as per my wider comment.

chrismytton committed:
Document ocd_ids_from_csv function

tmtmtmtm commented:
I'm not sure if it's worth fixing up here, as we're probably into diminishing returns, but certainly when we _do_ extract this to a library, the `ocd_ids_from_csv` should do a lot less. At the minute it's responsible for at least three distinct things:
1. parsing the CSV data provided
2. remapping the columns in that, if required
3. turning each row into a OCD hash

I think it would be better to have a separate function that turns a (correctly keyed) single hash of the different area levels into the OCD data (which can then be independently called, tested etc), and have a different function responsible for extracting and converting each of the rows that will be passed to that.

chrismytton committed:
Create separate function for checking if cell is empty

Rather than have a special case for '-' and '_' in the constituency
column I've extracted the check into a separate method which can be
called for any cell.

chrismytton commented:
I've opened https://github.com/theyworkforyou/uganda_ocd_ids/issues/2 to track the wider refactor, so merging this now.
